This notebook shows how the [Online Retail Data Set](http://archive.ics.uci.edu/ml/datasets/online+retail) from UCI was processed to be used on the MBA Tutorial.

In [1]:
import pandas as pd
# pd.set_option('display.max_rows', None)
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Obtain data directly from website
df = pd.read_excel("http://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx")

In [3]:
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [5]:
len(df['Description'].unique())

4224

In [6]:
# entries per country
df['Country'].value_counts()

United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

In [7]:
# list with countries names
countries_list = df['Country'].unique().tolist()

In [8]:
# Unique items per country
unique_items_list = [len(df[df['Country']==country]['Description'].unique()) for country in countries_list]
# unique_items_list

In [9]:
# unique number of transactions (InvoiceId) per country
unique_invoiceNo_list = [len(df[df['Country']==country]['InvoiceNo'].unique()) for country in countries_list]
# unique_invoiceNo_list

In [10]:
# Dataframe with number of transactions and unique items per country
df_info = pd.DataFrame({'Country':countries_list, "Number unique InvoiceId":unique_invoiceNo_list, 'Number unique items':unique_items_list})

How many unique invoices and unique items per country?

In [11]:
df_info.sort_values(['Number unique InvoiceId','Number unique items'])

Country  Number unique InvoiceId  Number unique items
33                Brazil                        1                   32
27               Lebanon                        1                   45
37                   RSA                        1                   58
29          Saudi Arabia                        2                    9
28  United Arab Emirates                        3                   68
23               Bahrain                        4                   16
13             Lithuania                        4                   29
30        Czech Republic                        5                   25
35    European Community                        5                   50
24                Greece                        6                  138
31                Canada                        6                  147
15               Iceland                        7                  103
34                   USA                        7                  163
21                Israel                        9                  257
36                 Malta                       10                   99
26             Singapore                       10                  179
32           Unspecified                       13                  344
25             Hong Kong                       15                  194
20               Austria                       19                  307
18                Cyprus                       20                  498
17               Denmark                       21                  252
9                 Poland                       24                  205
14                 Japan                       28                  217
16       Channel Islands                       33                  436
5                 Norway                       40                  595
19                Sweden                       46                  261
22               Finland                       48                  465
11                 Italy                       55                  480
2              Australia                       69                  609
10              Portugal                       71                  714
7            Switzerland                       74                  987
3            Netherlands                      101                  805
8                  Spain                      105                 1107
12               Belgium                      119                  785
6                   EIRE                      360                 2027
1                 France                      461                 1565
4                Germany                      603                 1703
0         United Kingdom                    23494                 4203

# Creating a Dataset with Products Sold in the Netherlands

Netherlands has 101 transactions and 805 unique items.

In [12]:
df_NL = df[df['Country']=='Netherlands']
df_NL.head()

InvoiceNo StockCode                        Description  Quantity  \
385      536403     22867            HAND WARMER BIRD DESIGN        96   
386      536403      POST                            POSTAGE         1   
37952    539491     21981       PACK OF 12 WOODLAND TISSUES         12   
37953    539491     21986   PACK OF 12 PINK POLKADOT TISSUES        12   
37954    539491     22720  SET OF 3 CAKE TINS PANTRY DESIGN          2   

              InvoiceDate  UnitPrice  CustomerID      Country  
385   2010-12-01 11:27:00       1.85     12791.0  Netherlands  
386   2010-12-01 11:27:00      15.00     12791.0  Netherlands  
37952 2010-12-20 10:09:00       0.29     14646.0  Netherlands  
37953 2010-12-20 10:09:00       0.29     14646.0  Netherlands  
37954 2010-12-20 10:09:00       4.95     14646.0  Netherlands

# Creating Products Categories

In one part of the tutorial we create categories of products to helping simplifying the MBA process. Bellow we show how we access the list of unique products of the category `Vintage`.

In [13]:
# Items containing the word vintage in the description
df_NL['Description'][df_NL['Description'].str.contains('VINTAGE')]

57420         VINTAGE CARAVAN GREETING CARD 
57469       PLASTERS IN TIN VINTAGE PAISLEY 
63458         JUMBO BAG PINK VINTAGE PAISLEY
66022        PINK VINTAGE PAISLEY PICNIC BAG
66084       PLASTERS IN TIN VINTAGE PAISLEY 
                         ...                
534893    VINTAGE HEADS AND TAILS CARD GAME 
534913           PAPERWEIGHT VINTAGE COLLAGE
534928             VINTAGE DONKEY TAIL GAME 
534932           PAPER BUNTING VINTAGE PARTY
534949              LUNCH BAG VINTAGE DOILY 
Name: Description, Length: 87, dtype: object

In [14]:
list_vintage = df_NL['Description'][df_NL['Description'].str.contains('VINTAGE')].tolist()
list_vintage = [item.lower() for item in list_vintage]
list_vintage = list(set(list_vintage))
list_vintage

['playing cards vintage doiley ',
 'pink vintage spot beaker',
 'vintage doily jumbo bag red ',
 'paperweight vintage collage',
 'vintage heads and tails card game ',
 'mini paint set vintage ',
 'jumbo bag pink vintage paisley',
 'vintage doily travel sewing kit',
 'box of vintage alphabet blocks',
 'plasters in tin vintage paisley ',
 'airline bag vintage jet set red',
 '3 raffia ribbons vintage christmas',
 'set 12 vintage doily chalk ',
 'pack of 12 vintage doily tissues',
 '60 cake cases vintage christmas',
 'red vintage spot beaker',
 'boys vintage tin seaside bucket',
 'wrap  vintage doiley ',
 'lunch bag vintage doily ',
 'vintage doily deluxe sewing kit ',
 'paper bunting vintage party',
 'set of 12 t-lights vintage doiley',
 'jumbo bag vintage christmas ',
 'charlotte bag vintage alphabet ',
 'magnets pack of 4 vintage collage',
 'box of 6 mini vintage crackers',
 'pink vintage paisley picnic bag',
 'vintage caravan greeting card ',
 'girls vintage tin seaside bucket',
 'set 

In [15]:
print('There are {} unique products in the category Vintage'.format(len(list_vintage)))

There are 41 unique products in the category Vintage


Now back in creating the dataset we use for the tutorial...

In [16]:
# remove leading and trailing whitespaces.

df_NL['Description'] = df_NL['Description'].apply(lambda x: x.strip().lower())

# group by invoice and generate dataframe with transactions id and items 
df_NL = df_NL.groupby('InvoiceNo')['Description'].apply(lambda x: ','.join(x)).reset_index()
df_NL

# rename columns
df_NL.columns = ['TransactionId', 'Transaction']

# save in csv
df_NL.to_csv(r"C:\MKB_datalab\Tutorials\marketing-analysis-market-basket-analysis\data\raw\retailer_nl.csv", index = False)

In [17]:
pd.read_csv(r"C:\MKB_datalab\Tutorials\marketing-analysis-market-basket-analysis\data\raw\retailer_nl.csv")


TransactionId                                        Transaction
0          536403                    hand warmer bird design,postage
1          539491  pack of 12 woodland tissues,pack of 12 pink po...
2          539731  pack of 72 retrospot cake cases,easter tin kee...
3          541206  pack of 12 pink polkadot tissues,rose cottage ...
4          541570  strawberry lunch box with cutlery,dinosaur lun...
..            ...                                                ...
96        C543388                                            postage
97        C556117                 ass floral print multi screwdriver
98        C563593                                           discount
99        C564899  manual,36 foil heart cake cases,rain poncho re...
100       C577481                                            postage

[101 rows x 2 columns]